src/rust/ensogl/lib/text/src/buffer/view.rs

```diff
@@ -140,37 +191,165 @@ impl Default for ViewBuffer {
+
+    fn new_cursor(&self, location:Location) -> Selection {
+        let id = self.next_selection_id.get();
+        self.next_selection_id.set(id+1);
+        Selection::new_cursor(location,id)
+    }
+
+    fn add_cursor(&self, location:Location) -> selection::Group {
+        let mut selection = self.selection.borrow().clone();
+        let new_selection = self.new_cursor(location);
+        selection.merge(new_selection);
+        selection
+    }
+
+    fn set_newest_selection_end(&self, location:Location) -> selection::Group {
+        let mut group = self.selection.borrow().clone();
+        group.newest_mut().for_each(|s| s.end=location);
+        group
     }
 
+    fn set_oldest_selection_end(&self, location:Location) -> selection::Group {
+        let mut group = self.selection.borrow().clone();
+        group.oldest_mut().for_each(|s| s.end=location);
+        group
+    }
+
+    // FIXME: this should be made private in the future PRs.
     /// Insert new text in the place of current selections / cursors.
-    pub fn insert(&self, text:impl Into<Data>) {
-        let text = text.into();
-        for selection in &*self.selection.borrow() {
-            self.buffer.data.borrow_mut().insert(selection.range(),&text);
+    pub fn insert(&self, text:impl Into<Text>) -> selection::Group {
+        self.modify(Transform::LeftSelectionBorder,text)
+    }
+
+    fn delete_left(&self) -> selection::Group {
+        self.modify(Transform::Left,"")
+    }
+
+    /// Generic buffer modify utility. First, it transforms all selections with the provided
+    /// `transform`, and then it replaces the resulting selection diff with the provided `text`.
+    /// See its usages across the file to learn more.
+    ///
+    /// ## Implementation details.
+    /// This function converts all selections to byte-based ones first, and then applies all
+    /// modification rules. This way, it can work in an 1D byte-based space (as opposed to 2D
+    /// location-based space), which makes handling multiple cursors much easier.
+    fn modify(&self, transform:Transform, text:impl Into<Text>) -> selection::Group {
+        self.commit_history();
+        let text                    = text.into();
+        let text_byte_size          = text.byte_size();
+        let mut new_selection_group = selection::Group::new();
+        let mut byte_offset         = 0.bytes();
+        for rel_byte_selection in self.byte_selections() {
+            let byte_selection     = rel_byte_selection.map(|t|t+byte_offset);
+            let selection          = self.to_location_selection(byte_selection);
+            let new_selection      = self.moved_selection_region(transform,selection,false);
+            let new_byte_selection = self.to_bytes_selection(new_selection);
+            let byte_range         = range_between(byte_selection,new_byte_selection);
+            byte_offset           += text_byte_size - byte_range.size();
+            self.buffer.replace(byte_range,&text);
```

Comment:
There are 2 possible pasting modes. If you select a text like "abc" and you try to paste it in multiple selections, the WHOLE TEXT should be pasted in EACH selection - test it in Intellij, atom, vscode, everywhere. However, if you copied in multi-select mode, then you remember each selection as a separate component and then pasting should paste them separately. We do not have the later mode implemented here. Was the old text editor working this way exactly? If you copy-pasted a single string, was it pasting it in all selected places? It should.

Reply:
I pu the reference in my comment, but I can also paste the essence for you:
```rust
let trimmed    = text.trim_end_matches('\n');
        let cursor_ids = self.with_borrowed(|this| this.cursors.sorted_cursor_indices());
        // When we insert (e.g. paste) many lines in multicursor mode, under some circumstances
        // we insert one line per cursor, instead of having all cursors inserting the whole
        // content. Such situation we call here Line Per Cursor Edit.
        let is_line_per_cursor_edit = trimmed.contains('\n') && cursor_ids.len() > 1;

        if is_line_per_cursor_edit {
            let cursor_with_line = cursor_ids.iter().cloned().zip(trimmed.split('\n'));
            self.write_per_cursor(cursor_with_line);
        } else {
            let cursor_with_line = cursor_ids.iter().map(|cursor_id| (*cursor_id,text));
            self.write_per_cursor(cursor_with_line);
        };
```
It's not _exactly_ as it work in intellij (intellij makes distinction when copying text in multiline mode between "newlines", and "newcursors"), but pasting multiline text copied from gedit works this way.